modepy/modes.py

```diff
@@ -25,15 +25,30 @@
 from warnings import warn
 from math import sqrt
 from functools import singledispatch, partial
+from typing import Callable, Optional, Sequence, TypeVar, Tuple, Union, TYPE_CHECKING
 
 import numpy as np
 
 from modepy.spaces import FunctionSpace, PN, QN
 from modepy.shapes import Shape, Simplex, Hypercube
 
+if TYPE_CHECKING:
+    import pymbolic.primitives
+
+InputType = Union[np.ndarray, "pymbolic.primitives.Expression", float]
+InputTypeT = TypeVar("InputTypeT",
+        np.ndarray, "pymbolic.primitives.Expression", float)
+
+
 __doc__ = """This functionality provides sets of basis functions for the
 reference elements in :mod:`modepy.shapes`.
 
+.. class:: InputTypeT
```

Comment:
Not sure I'm loving `InputTypeT` as a name, particularly because it occurs as output. :) `RealT`? `CoordinateT`? `BasisValueT`?

Reply:
Renamed it to `RealValueT` in 1660f8e. Is that ok?

Reply:
Yep, I like that. Works for me!